src/lib/agent/models/MessageContext.ts

```diff
@@ -0,0 +1,29 @@
+import { AgentMessage } from '../AgentMessage';
+import { ConnectionRecord } from '../../storage/ConnectionRecord';
+
+export interface MessageContextParams {
+  connection?: ConnectionRecord;
+  senderVerkey?: Verkey;
+  recipientVerkey?: Verkey;
+}
+
+export class MessageContext<T extends AgentMessage = AgentMessage> {
+  message: T;
+  connection?: ConnectionRecord;
+  senderVerkey?: Verkey;
+  recipientVerkey?: Verkey;
+
+  constructor(message: T, context: MessageContextParams = {}) {
+    this.message = message;
+    this.recipientVerkey = context.recipientVerkey;
+
+    if (context.connection) {
+      this.connection = context.connection;
+      // TODO: which senderkey should we prioritize
+      // Or should we throw an error when they don't match?
+      this.senderVerkey = context.connection.theirKey || context.senderVerkey || undefined;
+    } else if (context.senderVerkey) {
+      this.senderVerkey = context.senderVerkey;
+    }
+  }
+}
```

Comment:
Yeah that is a good idea. It can serve as a general purpose class to add extra information to a message.

For the notation: let's discuss on the call